# Random Survival Forest — non-linear effects + AE burden

**Design Spec §5.2.** Second of the three-stage analytics progression. Adds two things Cox PH
(`6 - cox_ph.ipynb`, C-index 0.555) structurally can't capture: non-linear covariate effects, and
adverse-event burden as a feature (from `stg_ae`, row-level AE data staged but unused until now).

**`trial_id` as a covariate, not a hard stratum:** unlike the Cox model, RSF gets `trial_id`
one-hot encoded as an ordinary feature per Design Spec §5.2 ("trial_id included as a
covariate/cluster feature"). Tree-based models don't have the Cox model's singular-Hessian
problem with a covariate that's constant within a group — a tree can simply split on `trial_id`
directly wherever it's informative, so `kras_status` (excluded from the Cox model for exactly this
reason) can be included here.

**Validation: held-out trial**, per Design Spec §9. Train on PACCE + FOLFIRI (1,788 patients),
test on FOLFOX/PRIME held out entirely (935 patients) — the stronger test of whether the model
generalizes to a genuinely different patient population/protocol, rather than just interpolating
within trials it's already seen.

In [1]:
import numpy as np
import pandas as pd
from sqlalchemy import create_engine

pd.set_option('display.max_columns', None)

engine = create_engine(
    "postgresql+psycopg2://VincenzoVantornout:localdevpassword@localhost:5432/postgres"
)

patients = pd.read_sql("select * from public.fct_patient_survival", engine)
ae = pd.read_sql("select * from public.stg_ae", engine)

print(f"{len(patients)} patients, {len(ae)} adverse event rows")

2723 patients, 66723 adverse event rows


## AE burden — aggregating `stg_ae` from event-level to patient-level

`stg_ae` is one row per adverse event, not per patient (see `learning-notes/dbt.md`). RSF needs
one row per patient, so this aggregates event-level rows into a per-patient burden summary before
joining to `fct_patient_survival`:
- `n_ae_events` — total adverse events recorded
- `max_severity_grade` — worst CTCAE-style grade (1-5) reached
- `n_serious_ae` — count where `serious == 'Y'`
- `n_related_ae` — count where `related_to_drug == 'Y'` (causally linked to study drug, not
  disease progression or unrelated illness)

Patients with **zero** AE rows (if any) should get `0`/`NaN`→`0` here, not silently disappear from
a `groupby` — handled via a left join back onto the full patient list below, with burden columns
`fillna(0)` afterward.

In [2]:
ae_burden = ae.groupby(['trial_id', 'subjid']).agg(
    n_ae_events=('severity_grade', 'size'),
    max_severity_grade=('severity_grade', 'max'),
    n_serious_ae=('serious', lambda s: (s == 'Y').sum()),
    n_related_ae=('related_to_drug', lambda s: (s == 'Y').sum()),
).reset_index()

merged = patients.merge(ae_burden, on=['trial_id', 'subjid'], how='left')
burden_cols = ['n_ae_events', 'max_severity_grade', 'n_serious_ae', 'n_related_ae']

missing_ae = merged['n_ae_events'].isna().sum()
print(f"{missing_ae} patients with zero recorded AE rows (filled to 0)")

merged[burden_cols] = merged[burden_cols].fillna(0)
merged.head()

22 patients with zero recorded AE rows (filled to 0)


,trial_id,subjid,age,sex,race_category,ecog_baseline,diagnosis_type,weight_kg,height_cm,bsa_m2,kras_status,on_panitumumab,discontinuation_reason,discontinuation_reason_code,discontinuation_day,event,time_days,n_ae_events,max_severity_grade,n_serious_ae,n_related_ae
0,NCT00115225,249101001,67.0,Female,Black or African American,Fully active,Colon,73.4,167.60,1.848561,Wild-type,True,Other,88.0,685.0,True,668.0,36.0,4.0,2.0,10.0
1,NCT00115225,249101002,72.0,Male,White or Caucasian,Symptoms but ambulatory,Colon,77.5,182.90,1.984296,Mutant,False,,NaN,NaN,True,315.0,13.0,3.0,0.0,0.0
2,NCT00115225,249101003,78.0,Male,White or Caucasian,Fully active,Colon,89.0,184.66,2.136634,Failed,True,Adverse event,5.0,191.0,True,196.0,31.0,5.0,10.0,6.0
3,NCT00115225,249101005,66.0,Female,Black or African American,Symptoms but ambulatory,Colon,52.7,160.00,1.530432,Wild-type,False,,NaN,NaN,True,98.0,12.0,5.0,3.0,0.0
4,NCT00115225,249101006,67.0,Male,White or Caucasian,Fully active,Rectal,95.9,180.30,2.191573,Failed,True,Other,88.0,344.0,True,348.0,23.0,4.0,4.0,1.0


## Feature prep

Same `bsa_m2`-missing handling as the Cox notebook (3 patients dropped, count printed). Unlike
Cox, `kras_status` (with nulls as an explicit `"Unknown"` category) and `trial_id` are both
included as ordinary one-hot-encoded features — no stratification-induced singularity here, since
RSF splits on features directly rather than inverting a Hessian.

In [3]:
model_df = merged.copy()
model_df['kras_status'] = model_df['kras_status'].fillna('Unknown')

before = len(model_df)
model_df = model_df.dropna(subset=['bsa_m2'])
print(f"Dropped {before - len(model_df)} patients missing bsa_m2 ({len(model_df)} remaining)")

feature_cols = [
    'age', 'bsa_m2', 'on_panitumumab',
    'sex', 'race_category', 'ecog_baseline', 'diagnosis_type', 'kras_status',
    'trial_id',
] + burden_cols
categorical_cols = ['sex', 'race_category', 'ecog_baseline', 'diagnosis_type', 'kras_status', 'trial_id']

X = pd.get_dummies(model_df[feature_cols], columns=categorical_cols, drop_first=True)
X['on_panitumumab'] = X['on_panitumumab'].astype(int)

y_time = model_df['time_days'].reset_index(drop=True)
y_event = model_df['event'].astype(bool).reset_index(drop=True)
trial_id = model_df['trial_id'].reset_index(drop=True)
X = X.reset_index(drop=True)

print(X.shape)
X.head()

Dropped 3 patients missing bsa_m2 (2720 remaining)
(2720, 19)


,age,bsa_m2,on_panitumumab,n_ae_events,max_severity_grade,n_serious_ae,n_related_ae,sex_Male,race_category_Other,race_category_White or Caucasian,ecog_baseline_In bed less than 50% of the time,ecog_baseline_Symptoms but ambulatory,diagnosis_type_Rectal,kras_status_Failed,kras_status_Mutant,kras_status_Unknown,kras_status_Wild-type,trial_id_NCT00339183,trial_id_NCT00364013
0,67.0,1.848561,1,36.0,4.0,2.0,10.0,False,False,False,False,False,False,False,False,False,True,False,False
1,72.0,1.984296,0,13.0,3.0,0.0,0.0,True,False,True,False,True,False,False,True,False,False,False,False
2,78.0,2.136634,1,31.0,5.0,10.0,6.0,True,False,True,False,False,False,True,False,False,False,False,False
3,66.0,1.530432,0,12.0,5.0,3.0,0.0,False,False,False,False,True,False,False,False,False,True,False,False
4,67.0,2.191573,1,23.0,4.0,4.0,1.0,True,False,True,False,False,True,True,False,False,False,False,False


## Held-out trial split

FOLFOX/PRIME (`NCT00364013`) held out entirely for testing; PACCE + FOLFIRI used for training —
per Design Spec §9 and the earlier scoping decision.

In [4]:
from sksurv.util import Surv

HELD_OUT_TRIAL = 'NCT00364013'

train_mask = trial_id != HELD_OUT_TRIAL
test_mask = trial_id == HELD_OUT_TRIAL

X_train, X_test = X[train_mask], X[test_mask]
y_train = Surv.from_arrays(event=y_event[train_mask], time=y_time[train_mask])
y_test = Surv.from_arrays(event=y_event[test_mask], time=y_time[test_mask])

print(f"Train: {len(X_train)} patients (PACCE + FOLFIRI)")
print(f"Test:  {len(X_test)} patients (FOLFOX/PRIME, fully held out)")

Train: 1786 patients (PACCE + FOLFIRI)
Test:  934 patients (FOLFOX/PRIME, fully held out)


## Fit the Random Survival Forest

`min_samples_leaf=15` avoids overfitting to tiny leaf groups given ~1,800 training patients;
`n_estimators=200` for a stable forest without an excessive fit time on this dataset size.

In [5]:
from sksurv.ensemble import RandomSurvivalForest

rsf = RandomSurvivalForest(
    n_estimators=200,
    min_samples_leaf=15,
    n_jobs=-1,
    random_state=42,
)
rsf.fit(X_train, y_train)
print("Fit complete.")

Fit complete.


## Evaluate on the held-out trial

`rsf.score()` computes Harrell's concordance index directly on held-out data — same metric as the
Cox model's `concordance_index_`, so the two numbers are directly comparable. This is a genuine
generalization test, not an in-sample number: the forest never saw a single FOLFOX/PRIME patient
during training.

In [6]:
train_cindex = rsf.score(X_train, y_train)
test_cindex = rsf.score(X_test, y_test)

print(f"Train C-index: {train_cindex:.3f}")
print(f"Held-out (FOLFOX/PRIME) C-index: {test_cindex:.3f}")
print(f"Cox PH baseline (in-sample, stratified): 0.555")

Train C-index: 0.748
Held-out (FOLFOX/PRIME) C-index: 0.609
Cox PH baseline (in-sample, stratified): 0.555


## Feature importance — did AE burden actually help?

`RandomSurvivalForest` doesn't expose `feature_importances_` the way a standard sklearn
`RandomForestClassifier` does (impurity isn't directly comparable across log-rank splits) — the
standard approach is **permutation importance**: shuffle one feature at a time on the held-out
set and measure how much the C-index drops. A feature that matters a lot will hurt the score badly
when scrambled; a useless one won't move it. This is the direct answer to whether staging `stg_ae`
and building the AE-burden aggregation earlier in this notebook was actually worth it.

In [7]:
from sklearn.inspection import permutation_importance

perm = permutation_importance(rsf, X_test, y_test, n_repeats=15, random_state=42, n_jobs=-1)

importances = pd.DataFrame({
    'feature': X_test.columns,
    'importance_mean': perm.importances_mean,
    'importance_std': perm.importances_std,
}).sort_values('importance_mean', ascending=False)

importances.head(15)

,feature,importance_mean,importance_std
3,n_ae_events,0.036692,0.007788
4,max_severity_grade,0.027620,0.004840
11,ecog_baseline_Symptoms but ambulatory,0.012246,0.004989
10,ecog_baseline_In bed less than 50% of the time,0.012198,0.003104
5,n_serious_ae,0.007998,0.002717
6,n_related_ae,0.007039,0.003193
0,age,0.005074,0.002172
2,on_panitumumab,0.003672,0.002818
1,bsa_m2,0.000992,0.001594
7,sex_Male,0.000513,0.000736


## Findings

**Held-out C-index: 0.609**, vs. Cox PH's 0.555 — a real improvement on a genuine
generalization test (FOLFOX/PRIME never seen during training), not an in-sample number inflated by
memorization. This is the comparison Design Spec §5.2/§9 asks for: RSF should improve on the
linear baseline by capturing non-linear effects and AE burden, and it does.

**Train (0.748) vs. held-out (0.609) gap** indicates some overfitting to the training trials, as
expected for a tree ensemble vs. a linear model — worth keeping in mind rather than over-trusting
the train number, which is why the held-out score, not the train score, is the one to report.

**AE burden was the single biggest reason RSF beat Cox.** `n_ae_events` and `max_severity_grade`
are the top 2 most important features by permutation importance — directly validating the earlier
decision to stage `stg_ae` and build the per-patient aggregation in this notebook. ECOG baseline
status (the dominant Cox PH driver) still matters here too, ranking 3rd/4th.

**Caution on `n_ae_events` — a real methodological point, not just a modeling footnote:** more
adverse events can simply mean *more time on study to accumulate them*, not necessarily higher
biological risk — a patient followed for 800 days has more opportunity to rack up AEs than one
followed for 50, independent of how sick they actually are. This is a length-of-follow-up
confound, not a clean causal signal. Fine for this predictive-only stage (Design Spec §5.2 doesn't
require causal AE effects), but this is exactly the kind of confound the TMLE stage (§5.3) needs
to handle explicitly and carefully for whichever modifiable driver gets chosen there — flagging it
now so it isn't rediscovered from scratch later.

**`trial_id` dummies and `kras_status` show ~zero importance** — expected for `trial_id`, since
the held-out trial's dummy column was constant (all-0 or all-1) throughout training and testing
respectively, giving the forest nothing to split on; the model's real generalization signal comes
entirely from the other covariates, which is exactly what a held-out-trial test is supposed to
verify. `kras_status`'s ~zero importance here is consistent with Cox PH's earlier structural
finding that it's tightly confounded with trial membership.

**Next**: TMLE (Design Spec §5.3) — the causal-inference stage, estimating the effect of one
specific modifiable dropout driver. Needs an explicit ICH E9(R1) estimand statement before fitting
anything, per the spec.